### DISCUSSION TITLE
Making it possible to set client directives dynamically

### DISCUSSION THREAD

**Discussion by manuelschroederdev** [withastro/roadmap#431] [Proposal]
Hey all,

I would like to propose the feature to set client directives dynamically. Especially when working with a CMS, it would be great to give content editors the possibility to choose with what priority a JS component should be hydrated. Technically, this is already possible right now, but it doesn't really in very clean and DRY code.

Here's an example of how I would currently approach this (using Storyblok as a headless CMS):

```javascript
---
import DynamicIsland from './DynamicIsland.svelte'

const { blok } = Astro.props
---
{blok.directive === 'visible' && <DynamicIsland blok={blok} client:visible />}
{blok.directive === 'load' && <DynamicIsland blok={blok} client:load />}
{blok.directive === 'idle' && <DynamicIsland blok={blok} client:idle />}
```

Being able to set the directive dynamically would result in shorter, more elegant code, somewhat similar to this example:

```javascript
---
import DynamicIsland from './DynamicIsland.svelte'

const { blok } = Astro.props
---
<DynamicIsland blok={blok} client={blok.directive} />
```

I already discussed this with @bholmesdev, and the challenge would be that the values would have to be known at compile-time.

Possible solutions could be:

- assuming that all directive resources are needed whenever a variable is used
- defining in `astro.config.mjs` which resources should always be loaded
- defining on a component level which resources should always be loaded

I'd love to hear your thoughts. 🙂

**Comment by matthewp** [comment]
I don't think the values do need to be known at compile time. Those directives are inlined in the HTML and added dynamically when used. So I don't think there's a technical limitation here.

We do need to know that a component uses *some* client-side directive however. So your `client` idea would work, technically.

We already do have something like you describe, the compiler transforms `client:{directive}` to `client:component-hydration={directive}` here: https://github.com/withastro/compiler/blob/23d1fc05fc6d760708a238699f3ac739a458fa7b/internal/transform/transform.go#L342

This was never intended to be used by the user and probably won't work if you tried using it, but I don't see a reason why we couldn't make this an official, low-level way to add dynamic client directives.

**Comment by manuelschroederdev** [comment]
Thank you very much for the insightful reply, @matthewp. It's great to know that the basic functionality already exists under the hood. I'd love to see this being implemented for the users.

**Comment by sph-e** [comment]
Hi, I would love to see this become stable too @matthewp

**Comment by MoustaphaDev** [comment]
Would be exciting to have this! As @matthewp pointed out, the directives are passed as props to the components and the decision on which directive to load is made dynamically at runtime.
I'm gonna explore the idea and maybe draft up a proposal next week. Keep an eye out for it! 😊

**Comment by MoustaphaDev** [comment]
# Summary

Dynamically set client directives

# Background & Motivation

Currently, to set different client directives on a component in Astro, we have to write a lot of boilerplate code. This is because we have to check for the directive name and then render the component with the correct directive. This is even more cumbersome when we have multiple components that need to be rendered with different directives like in a CMS setting.

# Goals

-   Make it easier to set client directives dynamically
-   Reduce boilerplate code when dynamically setting client directives

# Example

Instead of writing this code:

```astro
---
import DynamicIsland from './DynamicIsland.svelte'

const { blok } = Astro.props
---
{blok.directive === 'visible' && <DynamicIsland blok={blok} client:visible />}
{blok.directive === 'load' && <DynamicIsland blok={blok} client:load />}
{blok.directive === 'idle' && <DynamicIsland blok={blok} client:idle />}
...
```

We can write this much cleaner and more concise code:

```astro
---
import DynamicIsland from './DynamicIsland.svelte'

const { blok } = Astro.props
---
<DynamicIsland blok={blok} client:params={{ directive: blok.directive }} />
```

**Reply by MoustaphaDev** [reply]
_[See the exploration branch](https://github.com/withastro/astro/compare/main...feat/dynamic-client-directives)_

**Reply by bholmesdev** [reply]
Ah, this is a clever idea to avoid compiler changes! I’m curious if @natemoo-re has thoughts. Would it be possible to add variable interpolation to directives? Or is this solution a good middle ground?

**Reply by natemoo-re** [reply]
This is extremely interesting! I need to sit with this idea for a bit, but I love that it works well with our existing model and indirectly opens up the ability to chain multiple directives together.

The only directive that this doesn't support is `client:only` because those imports need to be stripped during the build. So that's an open question for me.

I also wonder if it should be named `use:client`? Slightly different than our existing scheme, but aligns with what's going on in the ecosystem.

**Reply by MoustaphaDev** [reply]
This became suddenly much more complex than I thought, thanks for the heads-up about the `client:only` directive not working!

We could name it `use:client` but that will require some (seemingly minimal) compiler changes I think, but I can see the appeal.

Plus naming it `client:params` would be a breaking change, I think, as custom client directives were introduced (but that's not too bad, we could add this feature as a flag)

I'll explore possibilities to support the `client:only` directive and report back here.
